rd-net/Lifetimes/Serialization/UnsafeWriter.cs

```diff
@@ -102,8 +110,29 @@ public void Dispose()
     [MethodImpl(MethodImplAdvancedOptions.AggressiveInlining)]
     public static Cookie NewThreadLocalWriter()
     {
+      return NewThreadLocalWriterImpl(false);
+    }
+
+    [MethodImpl(MethodImplAdvancedOptions.AggressiveInlining)]
+    public static Cookie NewThreadLocalWriterWithCleanup()
+    {
+      return NewThreadLocalWriterImpl(true);
+    }
+
+    [MethodImpl(MethodImplAdvancedOptions.AggressiveInlining)]
+    private static Cookie NewThreadLocalWriterImpl(bool releaseThreadStaticWriterIfNewlyCreated)
```

Comment:
I guess the purpose of this method is to release 1MB of unsafe memory, when no cookies are on stack?

There are things to think about:

- Each time this method is used 1MB will be allocated and that is very time-consuming. Design of UnsafeWriter assumes that it's used on limited amount of threads with no penalty on creation new Cookie (except very first one). I guees this assumption was violated if you need this method. Maybe for short-living threads it's better to allocate less than 1Mb of memory?
- It's worth to deallocate unsafe memory instantly, not relying of finalizer ~UnsafeWriter then. Or you can play with weak handle to some object that holds 1Mb of memory, trying to pool it, not deallocating and allocating memory rapidly in case of frequest NewThreadLocalWriterWithCleanup() calls.

Reply:
Unfortunately the heatbeat in SocketWire uses system timer and its callback is executed in random thread. The telemetry that Microsoft has shows that it leads to 50+ MB leaks quite frequently. 
It looks like there would be a better fix for it to create an UnsafeWriter over a preallocated buffer for the cases like sending the Ping message

Reply:
I see, thanks for explanation. Continuations on random threads and thread-bound memory do not work together well. The initial design assumed other usage and cookie creation should be as fast as possible. Maybe it worth to redesign memory allocation strategy for UnsageWriter, it's easy to abuse this API.
I can suggest special ThreadStatic variable (called e.g. "ShouldRetainMemory" in UnsafeWriter) that must be explicitly set for thread. Only if it's set UnsafeWriter takes memory from thread-local. Otherwiese - from pool.

Reply:
This is more or less the idea we agreed on while discussing internally (we used the term "our" threads) and will implement the fix by the ReSharper/ Rider 2020.3 RTM.  Or you can do it yourself if you wish))

Reply:
I'd like so but It would be impossible for me to profile it on ReSharper and adjust my theoretical hypothesis according brutal reality. =)